### PR TITLE
Resolves #15 - Adopt better logcontext discipline

### LINKED ISSUE #15: database queries run from sentinel logcontext
we see these warnings and errors every 60 seconds:

```
2021-06-08 11:04:13,388 - synapse_spamcheck_badlist.bad_list_filter - 105 - INFO - sentinel - _update_links_automaton: fetching links from table image_filter.iwf_links
2021-06-08 11:04:13,388 - synapse.storage.database - 671 - WARNING - sentinel - Starting db txn 'Fetch links from the table' from sentinel context
2021-06-08 11:04:13,388 - synapse.logging.scopecontextmanager - 73 - ERROR - sentinel - Tried to activate scope outside of loggingcontext
2021-06-08 11:04:13,404 - synapse.storage.database - 720 - WARNING - sentinel - Starting db connection from sentinel context: metrics will be lost
```

The problem is that the updater runs as a timed task. We don't have great support for this in the module API currently (something that also came up in the context of https://github.com/matrix-org/synapse/pull/9884), so I'm going to try to update the module API to sort that out.

### PR DESCRIPTION
**Pull request by Yoric**
Will this do the trick?

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -1,13 +1,13 @@
-from setuptools import setup, find_packages
+from setuptools import find_packages, setup
 
 setup(
     name="synapse-spamcheck-badlist",
-    version="0.2.4",
+    version="0.2.7",
```
Comment (richvdh) [inline setup.py:5]: .oO (what happened to 5 and 6)

File: synapse_spamcheck_badlist/bad_list_filter.py [hunk 2]
```diff
@@ -85,16 +87,20 @@ def __init__(self, config, api):
...
-        self._update_links_loop.start(pull_from_db_every_sec, now=False)
+
         # As soon as we can, run the first fetch.
         # Note that we have no guarantee that this is finished
         # by the time we receive the first message, so we need
         # a fallback in `_get_links_automaton()`.
         reactor.callWhenRunning(
-            lambda: defer.ensureDeferred(self._update_links_automaton())
+            lambda: defer.ensureDeferred(
+                context.run_in_background(self._update_links_automaton)
```
Comment (richvdh) [inline synapse_spamcheck_badlist/bad_list_filter.py:102]: this ought to be `run_as_background_process`: `run_in_background` is only appropriate when there is an existing logcontext that will be active for the duration of the background task (ie, it's useful for parallelisation, but not so much for discrete tasks)

File: synapse_spamcheck_badlist/bad_list_filter.py [hunk 3]
```diff
@@ -24,6 +24,8 @@
 from twisted.internet.task import LoopingCall
 from twisted.internet.threads import deferToThread
 
+from synapse.logging import context
```
Comment (richvdh) [inline synapse_spamcheck_badlist/bad_list_filter.py:27]: <sigh> This isn't really part of synapse's public API, though I think it's the best bet for now.

@babolivier did we decide what to do about this? Should we just make things like `run_as_background_process` part of the public API? Or should we add things like `ModuleApi.call_at_startup` to solve individual usecases?

**Review by richvdh — APPROVED** [review]
looks pretty good to me. Can you use `run_as_background_process` instead though please.

File: synapse_spamcheck_badlist/bad_list_filter.py (hunk 3 above)
Comment (Yoric) [reply synapse_spamcheck_badlist/bad_list_filter.py:27]: As an API consumer, I'd find the latter more agreeable. It's weird to use methods from two different libraries depending on whether I want to run the code asap or later.

Comment (richvdh) [reply synapse_spamcheck_badlist/bad_list_filter.py:27]: that's fair, though I'm also concerned that we'll never cover all usecases and people will *still* end up having to use `run_as_background_process` directly.

Comment (babolivier) [reply synapse_spamcheck_badlist/bad_list_filter.py:27]: Can't we do both? I can see usecases for both of them.

**Pull request merged**